Add winrmexec on behalf of Ozelis (https://github.com/ozelis/winrmexec/blob/main/winrmexec.py)

Dfte opened the pull request:
This PR adds the winrmexec.py example (written by Ozelis https://github.com/ozelis/winrmexec/issues/2) to Impacket : 

<img width="905" height="219" alt="image" src="https://github.com/user-attachments/assets/175ead16-7fbc-43cf-88fd-28c9cd6e5c97" />

Full ReadMe available here https://github.com/ozelis/winrmexec/blob/main/README.md.

Note that the last comit I did switched datetime UTC to datetime.utc for retrocompatibility issue with python prior to 3.11. This is a WIP PR as few things will have to be modified.

Commit:
Add WinRMExec.py from Ozelis

Commit:
Switch UTC to timezone.utc for compatibility with python < 3.11

Dfte commented:
Hello @anadrianmanrique  @gabrielg5, I really want to get back to that PR but I'm kinda stuck because of the size of the code. How would you handle that case ? Make a impacket/impacket/winrm.py class that is then used by impacket/example/winrmexec.py ?

Or is it okay to have such a huge example directly ?

gabrielg5 commented:
Hi @Dfte,
Yes, splitting the example into different classes is the best way to go from here. Also, there are a couple duplicated functions there as well that would be better to reuse from the impacket library itself now. It'd be easier to maintain afterwards...

Let me shoot at it with AI if you want. Seems like something it could easily solve for us.
We can move forward from there

Dfte commented:
Yes go for that! Then I'll start working on it :P thx mate!

gabrielg5 committed:
Add shared WinRM WinRS support

- add reusable WinRM transports, WSMan helpers, and WinRS client support
- rewrite winrmexec as a thin Impacket-style example
- reuse shared WinRM builders in relay code
- centralize TLS server-end-point CBT helpers
- add focused WinRM tests

gabrielg5 commented:
Ok, let's continue from here.
I've done a quick review on those changes and tested basic auth with some commands / shell like version to see if at least those work - we'll need to do better testing with the other auth methods. And also check if all needed features are there

Main changes:
* added reusable `impacket.winrm` with WSMan builders/parsers, transports, and `WinRSClient`
* rewrote `examples/winrmexec.py` as a thin Impacket-style example with standard logging and `-shell-type {cmd,powershell}`
* deduplicated the WinRM WSMan request-building used by the relay code
* centralized the TLS server-end-point CBT helper
* added focused WinRM tests

Dfte commented:
Exceptional! I'll work on that PR this week end then :)! Hopefully we'll have this great feature fully tested soon!! Thanks dude!

Commit:
Add try except to prevent KRB SRV UNKNOWN PRINCIPAL

Commit:
Add white space between CMD > and PS >

Commit:
Add credits

Commit:
Removed unecessary imports

Commit:
Move try catch to winrmexec.py

Commit:
Add raise

Dfte commented:
Alright so here is a few testings:

* Password authentication works (http and https):

<img width="917" height="109" alt="image" src="https://github.com/user-attachments/assets/9a9da22e-38de-4112-963a-d3fa4d7b63d7" />

* Kerberos with CCACHE works (http and https):

<img width="1217" height="128" alt="image" src="https://github.com/user-attachments/assets/585c1013-dcee-4ce3-87c8-5ad0759a57c4" />

* Kerberos with AES key works (http and https):

<img width="1498" height="139" alt="image" src="https://github.com/user-attachments/assets/96f8c248-37fc-4600-9be0-68cd379ffe2a" />

* Basic authentication works on http (if Set-WSManInstance -ResourceURI winrm/config/service -ValueSet @{AllowUnencrypted=$true} is run before) and on https:

<img width="914" height="361" alt="image" src="https://github.com/user-attachments/assets/ceda0c72-43b8-4e68-832e-5c9223b442df" />

* NT authentication: there is a bug there:

<img width="899" height="93" alt="image" src="https://github.com/user-attachments/assets/5bd81f74-b5b2-4f48-a747-24b41275df0a" />

* Cert authentication: not tested yet.

About ntlmrelay, I have set the winrms endpoint to None in order to allow ntlm relay, looks good as well!

<img width="1006" height="518" alt="image" src="https://github.com/user-attachments/assets/eff188fc-c104-4dac-ad2a-2a3ae81c2962" />

Btw, once this PR is merged I'll move the CBT computing from TDS.py to the Tls.py library as well so that it can be reused as well as we mentionned before @anadrianmanrique :)!

Commit:
Patch NT hash authenticatino

Dfte commented:
Latest commit patches the NT hash authentication which was broken:

<img width="999" height="163" alt="image" src="https://github.com/user-attachments/assets/98236337-a10b-447b-98b5-4086f79ef6a8" />

Last thing to do is to check that cert-pem and cert-pass options work and add a way to pass a pfx file as well. So far I wasn't able to set up a lab to test these options, I always end up with the following 503 error when enabling the certificate authentication:

<img width="1040" height="92" alt="image" src="https://github.com/user-attachments/assets/477a6952-ba98-4c7c-9ceb-bb583bd1c879" />

Dfte commented:
Hey getting back to that PR. As far as I know the -cert*.
I'll take a look at why the -cert* options do not work correctly. Otherwise, the rest of the example script does work and we should be able to merge it :P

edcod3 commented:
Hey, I just came across this pull request while initially looking at WinRM kerberos support (for NetExec) and would love to see this merged!

In regard to the last comment about -cert* support in `winrmexec.py`, i tried testing the latest winrmexec.py from the `add_winrmexec` branch in my current AD lab setup and (at least for my setup) the client certificate authentication seemed to work just fine. I've also attached a screenshot below:

<img width="2217" height="1171" alt="screenshot" src="https://github.com/user-attachments/assets/fbdc9cfc-4ebe-4c9e-877e-7185f9d4c380" />

I've also tested the same functionality with evil-winrm and evil-winrm-py to double-check.

The one thing I modified was the winrm test in `tests/misc/test_winrm.py` for `test_keytab_rc4_hashes_are_reparsed_after_load`:
```python
self.assertEqual(captured['lmhash'], '')
self.assertEqual(captured['nthash'], '00112233445566778899AABBCCDDEEFF')
```
```python
self.assertEqual(captured['lmhash'], b'')
self.assertEqual(captured['nthash'], bytes.fromhex('00112233445566778899AABBCCDDEEFF'))
```
As the `nthash` and `lmhash` are decoded to bytes in `_parse_hashes`, this is the only change that had to be made to satisfy all test cases.

Thank you @Dfte and @gabrielg5 for your work on this. Hope this addition is of some help in your efforts to get this great work merged :)

Dfte commented:
Thank you for the review @edcod3 !! I have patched the test_winrm.py as mentionned!
@gabrielg5 let me know if you need anything else / have time to review or anything. I'll gladly work on this PR :)!
